src/Features/Core/Portable/ValidateRegexString/AbstractValidateRegexStringDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,102 @@
+
+namespace Microsoft.CodeAnalysis.ValidateRegexString
+{
+    internal abstract class AbstractValidateRegexStringDiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        private readonly int _stringLiteralKind;
+        private readonly ISyntaxFactsService _syntaxFacts;
+        private readonly ISemanticFactsService _semanticFacts;
+        private readonly IVirtualCharService _virtualCharService;
+
+        protected AbstractValidateRegexStringDiagnosticAnalyzer(
+            int stringLiteralKind,
+            ISyntaxFactsService syntaxFacts,
+            ISemanticFactsService semanticFacts,
+            IVirtualCharService virtualCharService)
+            : base(IDEDiagnosticIds.RegexPatternDiagnosticId,
+                   new LocalizableResourceString(nameof(FeaturesResources.Regex_issue_0), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+            _stringLiteralKind = stringLiteralKind;
+            _syntaxFacts = syntaxFacts;
+            _semanticFacts = semanticFacts;
+            _virtualCharService = virtualCharService;
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSemanticModelAction(AnalyzeSemanticModel);
+
+        private void AnalyzeSemanticModel(SemanticModelAnalysisContext context)
+        {
+            var semanticModel = context.SemanticModel;
+            var syntaxTree = semanticModel.SyntaxTree;
+            var cancellationToken = context.CancellationToken;
+            var options = context.Options;
+            var optionSet = options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var option = optionSet.GetOption(RegularExpressionsOptions.ReportInvalidRegexPatterns, syntaxTree.Options.Language);
+            if (!option)
+            {
+                return;
+            }
+
+            var detector = RegexPatternDetector.TryGetOrCreate(semanticModel, _syntaxFacts, _semanticFacts);
+            if (detector == null)
+            {
+                return;
+            }
+
+            var root = syntaxTree.GetRoot(cancellationToken);
+            Analyze(context, detector, _virtualCharService, root, cancellationToken);
+        }
+
+        private void Analyze(
+            SemanticModelAnalysisContext context, RegexPatternDetector detector,
+            IVirtualCharService virtualCharService, SyntaxNode node, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            foreach (var child in node.ChildNodesAndTokens())
```

Comment:
did you do this way since there was no RegisterSyntaxToken? we do have RegisterSyntaxNode. this basically visit every token except structure trivia

Reply:
I was concerned about all the overhead of calling back and forth.  Since each call would then need the callee side to then do a bunch of work all over again.  For example, we'd need to check options on every string token (and there can be a lot of those).  Similarly, we'd have to get the pattern detector set up each time.  I can switch to the other form.  But this seemed reasonable to me.